web/js/codeworld_shared.js

```diff
@@ -122,13 +122,111 @@ var hintBlacklist = [
     "Pi",
 ];
 
-var codeWorldBuiltinsDocs = {};
+
+// codeWorldDocs is variable containing annotations and documentation
+// of builtin and user-defined variables.
+// Expected format:
+// codeWorldDocs = {
+//   codeWorldLogo: {
+//     annotation: "Picture",
+//     doc: "The CodeWorld logo."
+//   }
+// }
+
+var codeWorldDocs = {};
+
+function renderHover(keyword) {
+    var elem = document.createElement('div')
+
+    if (!codeWorldDocs[keyword]){
+        return null;
+    };
+    elem.title = keyword;
+    var keywordData = codeWorldDocs[keyword];
+
+    // Variable name + annotaion
+    var topLine = document.createElement("div");
+    topLine.className = "hover-decl";
+    var wordElem = document.createElement("span");
+    wordElem.className = "hint-word";
+    wordElem.appendChild(document.createTextNode(keyword));
+    topLine.appendChild(wordElem);
+    if (keywordData.annotation) {
+        var annotation = document.createElement('div');
+        topLine.appendChild(document.createTextNode(" :: " + keywordData.annotation))
+    };
+    elem.appendChild(topLine)
+
+    // Docs
+    if (keywordData.doc) {
+        var docElem = document.createElement('div');
+        docElem.innerHTML = keywordData.doc;
+        docElem.className = "hover-doc";
+        elem.appendChild(docElem)
+    };
+    return elem;
+};
+
+function applyCodemirrorChange(line, change) {
+    if (change.origin === "+input"){
+        return line.slice(0, change.from.ch) + change.text.join('') + line.slice(change.from.ch);
+    } else if (change.origin === "+delete"){
+        return line.slice(0, change.from.ch) + line.slice(change.from.ch + 1);
+    } else if (["setValue", "complete"].indexOf(change.origin) != -1){
+        return;
+    } else {
+        console.log("Event \"" + change.origin + "\" not handled.");
+        return;
+    }
+}
+
+function getWordAndAnnotation(line) {
+    var annotation = null,
+        word = null;
+    // there is top-level assignment
+    if (/^\S+\s*=/.test(line)) {
+        word = line.split("=")[0].trim()
+    };
+    // there is top level type annotation
+    if (/^\S+\s*::/.test(line)) {
+        var splitted = line.split("::");
+        word = splitted[0].trim()
+        annotation = splitted[1].trim()
+    };
+    return {
+        word: word,
+        annotation: annotation
+    }
+}
+
+function updateDocs(cm, change) {
+    // take current line
+    var before = cm.getLine(change.from.line);
+    // make line after applied change
+    var after = applyCodemirrorChange(before, change);
+
+    before_data = getWordAndAnnotation(before)
+    after_data = getWordAndAnnotation(after)
+
+    // user updated name/annotaion of some variable
+    // updating hovers/hints information
+    if (!(before_data.word === after_data.word
+          && before_data.annotation === after_data.annotation
+          // We don't want to override builtins.
+          // Marker of builtin variable - it have docs.
```

Comment:
I'm not sure about this invariant that builtin symbols have docs.  It may be true today (though not always), but it could change in the future if we start to parse docs in the current module.  What about adding a boolean field to the object instead?

Reply:
Ok, will add.